compiler/rustc_passes/src/stability.rs

```diff
@@ -164,8 +167,27 @@ impl<'a, 'tcx> Annotator<'a, 'tcx> {
 
         let (stab, const_stab) = attr::find_stability(&self.tcx.sess, attrs, item_sp);
 
-        let const_stab = const_stab.map(|(const_stab, _)| {
+        let const_stab = const_stab.map(|(const_stab, const_span)| {
             let const_stab = self.tcx.intern_const_stability(const_stab);
+            if let Some(fn_sig) = fn_sig {
+                if !(fn_sig.header.constness == hir::Constness::Const
```

Comment:
I'm wondering whether we should call `is_const_fn_raw` instead if checking the sig. There are other ways to be const, like being in a const impl block. This could also make things easier for you as you can now just call a query on the def id instead of moving the fnsig around. Not sure if that will cause problems if called on non fn defids.

Reply:
I could pass a boolean `is_fn` to the `annotate` method and use `is_const_fn_raw` on the def id if it's the case? That would make it easier than having to play with the kind of node I am playing with.

Reply:
I just checked the body of `is_const_fn_raw`, it should not panic if called on non-functions.

Reply:
Mmmh, I still have to check for intrinsic (Rust & Platform) ABIs, right? So I should probably still pass the `fn_sig` through I think.. Unless `is_const_fn_raw` takes care of that in the background too?

Reply:
I tried this : 
```rust
let const_stab = self.tcx.intern_const_stability(const_stab);
let def_id = self.tcx.hir().local_def_id(hir_id).to_def_id();

if let Some(fn_sig) = fn_sig {
  if !(fn_sig.header.abi != Abi::RustIntrinsic
      && fn_sig.header.abi != Abi::PlatformIntrinsic
      && !self.tcx.is_const_fn_raw(def_id))
```

It makes everything const error out when it tries to build the `stage1` lib artifacts. Am I using this correctly? Or maybe `is_const_fn_raw` is only usable at a later stage?

I also noticed this: 
https://github.com/rust-lang/rust/blob/beac68a88711a90346ec8b68e3baefbec62b3b0d/src/librustc_passes/stability.rs#L86
`impl const Trait for Type` items forward their const stability to their immediate children`

Should I maybe use this route instead?

Reply:
My bad, I figured it out! I think I got a sweet implementation for this!

Reply:
looks great now!